src/Evolution/Particles/MonteCarlo/MonteCarloPacket.cpp

```diff
@@ -20,4 +20,22 @@ void Packet::renormalize_momentum(
       sqrt(momentum_upper_t) / get(lapse)[index_of_closest_grid_point];
 }
 
+double compute_fluid_frame_energy(
+    const Packet& packet, const Scalar<DataVector>& lorentz_factor,
+    const tnsr::i<DataVector, 3, Frame::Inertial>& lower_spatial_four_velocity,
+    const Scalar<DataVector>& lapse,
+    const tnsr::II<DataVector, 3, Frame::Inertial>& inv_spatial_metric) {
+  const size_t idx = packet.index_of_closest_grid_point;
+  double fluid_frame_energy =
+      get(lorentz_factor)[idx] * get(lapse)[idx] * packet.momentum_upper_t;
+  for (size_t i = 0; i < 3; i++) {
+    for (size_t j = 0; j < 3; j++) {
+      fluid_frame_energy -= inv_spatial_metric.get(i, j)[idx] *
+                            lower_spatial_four_velocity.get(i)[idx] *
+                            packet.momentum[idx];
```

Comment:
Should this by `j` instead of `idx`? Could you add a test to check for that? A few ideas that come to mind:
1. Set up a packet with `p^t=0` and `p_i={0.1, 0.35, 0.78}` and then check that you get the expected number out
2. Set up a packet with `p^t=0.9` and `p_i={0,0,0}` and then check you get the expected number out

In general just avoid 1 (or zero unless it's deliberate) since a multiplication bug isn't caught if you use 1 (2 is similarly bad, but slightly better). This is why we chose `p^t=0.9`

Reply:
Oh yeah, good catch...